商品出品機能

55Suape opened the pull request:
#What
商品出品機能の実装

#Why
商品出品機能の実装のため、以下ご確認をお願いします。

ログイン状態の場合は、商品出品ページへ遷移できる
https://gyazo.com/afaefa2b60e70d6ccd8022c43a2b3a5a

価格が入力されると同時に、販売手数料と販売利益が表示される
https://gyazo.com/c577001dd73762943de32e531c79649f

必須項目を適切に入力し「出品する」を押下すると、商品情報がデータベースに保存される
https://gyazo.com/8296b95ef3f3b1cec859066c2b9f0650

入力不備の状態で「出品する」を押下すると、情報は保存されず出品ページにエラー文言が表示される
https://gyazo.com/e36c0696b6c0d803321d73d202088b93

ログアウト状態で商品出品ページへ遷移しようとすると、ログインページ遷移
https://gyazo.com/2ae11f260d341d898b9c3052159b8716

Itemモデルの単体テスト結果
https://gyazo.com/c7ee7dce3e4919e90d0bfda6affec867

55Suape committed:
Itemモデルとitemsテーブルの作成

・Itemモデル作成
・itemsテーブル作成
・Itemモデルのバリデーション追記
・UserモデルとItemモデルのアソシエーション追記

55Suape committed:
ActiveHashの設定

55Suape committed:
Active Storageの導入

55Suape committed:
商品出品ページの表示＋Itemモデルの単体テスト

・ログアウト状態のユーザーが商品出品ページへ遷移しようとすると、ログインページへリダイレクト
・Itemモデルのバリデーション記述
・Itemモデルの単体テスト
（ActiveStorageの検証のため、FactoryBotにafterメソッドを用いて生成するダミーデータに画像を添付）

55Suape committed:
販売手数料・利益計算のJavaScript実装

item_price.jsで、views/items/new.html.erbのitem-priceフォームに入力した金額を取得しブラウザ上に描画する

minachanchan commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,5 @@
 class ApplicationController < ActionController::Base
+  before_action :authenticate_user!, except: [:index] # indexアクションの場合を除き、ログインしていないユーザーはログイン画面へ促す
```
```ruby
before_action :authenticate_user!, except: [:index]
```
こちらの記述はitems_controllerへ転記しましょう！
application_controller内へ記載した場合、全てのコントローラーで適用されるためでございます。

minachanchan commented on app/models/item.rb:
```diff
@@ -0,0 +1,23 @@
+  validates :schedule_id, numericality: { other_than: 1, message: "can't be blank" }
+  with_options presence: true do
+    validates :price, numericality: true, inclusion: { in: 300..9_999_999, allow_blank: true },
+                      format: { with: /\A[0-9]+\z/, allow_blank: true }
```
```ruby
format: { with: /\A[0-9]+\z/, allow_blank: true }
```
priceはinteger型のため正規表現を適用することができません。
半角整数値のみを許容するため、only_integerを使用しましょう！

minachanchan commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,73 @@
+        @item.category_id = ''
+        @item.valid?
+        expect(@item.errors.full_messages).to include("Category can't be blank")
+      end
```
active_hashで実装した5つの項目に関しては「---」が選択された場合は保存できない（idに1が選択された場合は保存できない）ことを確認するテストをそれぞれ実装しましょう！

minachanchan commented on spec/models/item_spec.rb:
```diff
@@ -0,0 +1,73 @@
+        @item.price = '200'
+        @item.valid?
+        expect(@item.errors.full_messages).to include('Price is not included in the list')
+      end
```
商品価格に関しては、以下を確認するテストコードをそれぞれ実装しましょう！
```
・10000000以上の値では保存できないこと
```

minachanchan replied:
以下について確認するテストコードをそれぞれ追記しましょう！

```
・画像が空では保存できないこと
・userが紐づいていないと保存できないこと
```

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いでございます。
また、トップページの出品ボタンへのリンクの実装がお済みでないようです。
こちらについてもご確認くださいませ。
次回レビュー依頼の際、出品ボタンからの画面遷移についてのキャプチャのご提示もお願いできればと存じます。

修正の後、再度レビューのご依頼頂けますようお願い申し上げます。

55Suape committed:
商品出品機能の修正（コントローラー）

before_action :authenticate_user!, except: [:index]の記述はitems_controllerへ移動。（application_controller内へ記載した場合、全てのコントローラーで適用されるため）

55Suape committed:
商品出品機能の修正（Itemモデル）

format: { with: /\A[0-9]+\z/, allow_blank: true }は正規表現となるが、priceはinteger(整数)型のため正規表現を適用することができない。そのため半角整数値のみを許容するonly_integerを使用。

55Suape committed:
商品出品機能の修正（Itemモデルのテスト項目追加）

spec/models/item_spec.rbに下記テスト項目を追加
・imageが空では登録できない
・ActiveHushの5項目は「1」の時登録できない
・priceは10000000以上の値では保存できない
・userが紐づいていないと保存できない

55Suape committed:
トップページ出品ボタンのリンク実装

<%= link_to 指定パス do %><% end %>で囲まれた中のものは、どれをクリックしても指定パスへ遷移する。

55Suape commented:
#What
商品出品機能の実装

#Why
いただいた修正事項を反映しましたので、ご確認をお願いします。

ログイン状態で出品ボタン押下後の遷移
https://gyazo.com/e5b13cf09740e86aeaaddb736dc9caa2

ログアウト状態で出品ボタン押下後の遷移
https://gyazo.com/023d8b5d7aeb30108c6b50f6d66785d7

minachanchan reviewed (commented):
レビューのご依頼ありがとうございます。

LGTMとさせていただきます。実装お疲れ様でした！
このブランチをマージし、masterブランチでプルしましょう。
機能が反映されていることを確認し、次の機能の実装に入りましょう！
引き続きの実装、応援しております！
